### PR TITLE
workflows: Add Code Scanning action, fix up failures

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/about-code-scanning
- I just joined the Code Scanning team at work and I figured I'd test out the actual product in the real world by seeing what things it points out for Homebrew, a reasonably large Ruby project.
- This adds a config file to exclude `Library/Homebrew/vendor` as we can't fix problems within gems. :-)

### CODE REVIEW

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+name: "CodeQL"
+
+on:
+ push:
+   branches: [ master ]
+ pull_request:
+   branches: [ master ]
+ schedule:
+   - cron: '30 2 * * 0'
```
Comment (MikeMcQuaid) [inline .github/workflows/codeql-analysis.yml:9]: Not sure about a scheduled run: who gets notified and who would fix these up?

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+   runs-on: ubuntu-latest
+   permissions:
+     actions: read
+     contents: read
+     security-events: write
+
+   strategy:
+     fail-fast: false
+     matrix:
+       language: [ 'ruby' ]
```
Comment (MikeMcQuaid) [inline .github/workflows/codeql-analysis.yml:23]: Seems like we don't need a matrix here?

File: Library/Homebrew/download_strategy.rb [hunk 3]
```diff
@@ -1404,18 +1404,18 @@ def self.detect_from_url(url)
     when %r{^https?://www\.apache\.org/dyn/closer\.cgi},
          %r{^https?://www\.apache\.org/dyn/closer\.lua}
       CurlApacheMirrorDownloadStrategy
-    when %r{^https?://(.+?\.)?googlecode\.com/svn},
+    when %r{^https?://([A-Za-z0-9\-.]+\.)?googlecode\.com/svn},
```
Comment (MikeMcQuaid) [inline Library/Homebrew/download_strategy.rb:1407]:
```suggestion
    when %r{^https?://(\w\-.]+\.)?googlecode\.com/svn},
```

maybe and elsewhere?

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (issyl0) [reply .github/workflows/codeql-analysis.yml:9]: This was the default when I clicked "add code scanning" in the UI. As below with the `matrix`. I'll remove them, you're right, a bit redundant.

File: Library/Homebrew/download_strategy.rb (hunk 3 above)
Comment (issyl0) [reply Library/Homebrew/download_strategy.rb:1407]: `\w` includes `A-Za-z0-9` but also `_` which [isn't valid in a hostname](https://man7.org/linux/man-pages/man7/hostname.7.html).

Comment (MikeMcQuaid) [reply Library/Homebrew/download_strategy.rb:1407]: TIL!

File: .github/workflows/codeql-analysis.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
+name: "CodeQL"
+
+on:
+ push:
+   branches: [ master ]
+ pull_request:
+   branches: [ master ]
```
Comment (MikeMcQuaid) [inline .github/workflows/codeql-analysis.yml:7]:
```suggestion
 push:
   branches:
     - master
 pull_request:
```

**Review by MikeMcQuaid — APPROVED** [review]

File: .github/workflows/codeql-analysis.yml (hunk 4 above)
Comment (issyl0) [reply .github/workflows/codeql-analysis.yml]: It's unhappy with this:

Screenshot 2022-04-18 at 13 26 03

Comment (MikeMcQuaid) [reply .github/workflows/codeql-analysis.yml]: Bah, ok, fair enough, add back a `- master` for `pull_request`.

Comment (issyl0) [reply .github/workflows/codeql-analysis.yml]: Done!

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request merged**